modules/nlu/src/backend/prediction-handler.ts

```diff
@@ -0,0 +1,64 @@
+import * as sdk from 'botpress/sdk'
+
+type ModelGetter = (lang: string) => Promise<sdk.NLU.Model | undefined>
+
+export class PredictionHandler {
+  constructor(private getLatestModel: ModelGetter, private engine: sdk.NLU.Engine, private defaultLanguage: string) {}
+
+  async predict(textInput: string, includedContexts: string[]) {
+    const defaultResults = await this.engine.predict(textInput, includedContexts, this.defaultLanguage)
+    const detectedLanguage = defaultResults.detectedLanguage
+    const detectedIsDefault = detectedLanguage === this.defaultLanguage
+
+    let nluResults = defaultResults
+    if (!detectedIsDefault) {
+      nluResults = await this.engine.predict(textInput, includedContexts, detectedLanguage)
+    }
+
+    if (this.isEmptyOrError(nluResults)) {
+      nluResults = await this._loadAndPredict(textInput, includedContexts, detectedLanguage)
+    }
+
+    if (!detectedIsDefault && this.isEmptyOrError(nluResults)) {
+      nluResults = await this._fallbackOnDefault(textInput, includedContexts, defaultResults)
+    }
+
+    if (this.isEmptyOrError(nluResults)) {
+      const msg = detectedIsDefault
+        ? `no model found for language (${this.defaultLanguage})`
+        : `no model found for either bot's default language (${this.defaultLanguage}) or detected language (${detectedLanguage})`
+      throw new Error(msg)
+    }
+
+    return nluResults
+  }
+
+  private isEmptyOrError(nluResults: sdk.IO.EventUnderstanding | undefined) {
+    return !nluResults || (nluResults.error && nluResults.error === 'invalid_predictor')
+  }
+
+  private async _loadAndPredict(
+    textInput: string,
+    includedContexts: string[],
+    language: string
+  ): Promise<sdk.IO.EventUnderstanding | undefined> {
+    const model = await this.getLatestModel(language)
```

Comment:
this works but it bugs me that getLatestModel is a function passed as constructor dependency. When reading I first thought that this was missing an implementation. Don't have a solution atm though